File: src/modules/user/controllers/user.controller.ts
import { $Enums } from '@prisma/client';
import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Patch,
  Post,
  Put,
  Query,
  UploadedFile,
  UseInterceptors,
} from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';
import { User } from 'src/shared/decorators/user.decorator';
import { UserCreateDto } from '../domain/dto/user-create.dto';
import { UserFindManyDto } from '../domain/dto/user-find-many.dto';
import { UserUpdateSelfDto } from '../domain/dto/user-update-self.dto';
import { UserUpdateStatusDto } from '../domain/dto/user-update-status.dto';
import { UserUpdateDto } from '../domain/dto/user-update.dto';
import { UserCreateService } from '../services/user-create.service';
import { UserDeleteService } from '../services/user-delete.service';
import { UserFindManyService } from '../services/user-find-many.service';
import { UserFindOneService } from '../services/user-find-one.service';
import { UserFindSelfService } from '../services/user-find-self.service';
import { UserRequest } from 'src/shared/types/user-request';
import { UserUpdateSelfService } from '../services/user-update-self.service';
import { UserUpdateStatusService } from '../services/user-update-status.service';
import { UserUpdateService } from '../services/user-update.service';
import { RequireAbilities } from 'src/shared/decorators/require-abilities.decorator';
import { FileInterceptor } from '@nestjs/platform-express';
import { UserUpdateAvatarService } from '../services/user-update-avatar.service';
import { PaginationResponseDto } from 'src/shared/domain/dto/pagination-response.dto';
import { UserEntity } from '../domain/entities/user.entity';
import { UserDeleteAvatarService } from '../services/user-delete-avatar.service';
import { IgnoreAbilities } from 'src/shared/decorators/ignore-abilities.decorator';

@ApiTags('Users')
@Controller('users')
export class UserController {
  constructor(
    private readonly userFindManyService: UserFindManyService,
    private readonly userFindOneService: UserFindOneService,
    private readonly userFindSelfService: UserFindSelfService,
    private readonly userCreateService: UserCreateService,
    private readonly userUpdateService: UserUpdateService,
    private readonly userUpdateSelfService: UserUpdateSelfService,
    private readonly userUpdateStatusService: UserUpdateStatusService,
    private readonly userDeleteService: UserDeleteService,
    private readonly userUpdateAvatarService: UserUpdateAvatarService,
    private readonly userDeleteAvatarService: UserDeleteAvatarService,
  ) {}

  @Get()
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.READ)
  async findAll(
    @Query() params: UserFindManyDto,
    @User() userRequest: UserRequest,
  ): Promise<PaginationResponseDto<UserEntity>> {
    return await this.userFindManyService.execute(params, userRequest);
  }

  @Get('self')
  @IgnoreAbilities()
  async findSelf(@User() userRequest: UserRequest): Promise<UserEntity> {
    return await this.userFindSelfService.execute(userRequest);
  }

  @Get(':id')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.READ)
  async findOne(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userFindOneService.execute(id, userRequest);
  }

  @Post()
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.CREATE)
  async create(
    @Body() dto: UserCreateDto,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userCreateService.execute(dto, userRequest);
  }

  @Put('self')
  @IgnoreAbilities()
  async updateSelf(
    @Body() dto: UserUpdateSelfDto,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userUpdateSelfService.execute(dto, userRequest);
  }

  @Post(':id/avatar')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.UPDATE)
  @UseInterceptors(FileInterceptor('file'))
  async uploadAvatar(
    @Param('id') id: number,
    @UploadedFile() file: Express.Multer.File,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userUpdateAvatarService.execute(id, file, userRequest);
  }

  @Delete(':id/avatar')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.DELETE)
  async deleteAvatar(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userDeleteAvatarService.execute(id, userRequest);
  }

  @Put(':id')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.UPDATE)
  async update(
    @Param('id') id: number,
    @Body() dto: UserUpdateDto,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userUpdateService.execute(id, dto, userRequest);
  }

  @Delete(':id')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.DELETE)
  async delete(
    @Param('id') id: number,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userDeleteService.execute(id, userRequest);
  }

  @Patch(':id/status')
  @RequireAbilities($Enums.AbilityCodes.USERS, $Enums.AbilityActions.UPDATE)
  async updateStatus(
    @Param('id') id: number,
    @Body() dto: UserUpdateStatusDto,
    @User() userRequest: UserRequest,
  ): Promise<UserEntity> {
    return await this.userUpdateStatusService.execute(id, dto, userRequest);
  }
}

File: src/modules/user/controllers/user_controller.py
from ninja import Router, Query, File
from ninja.files import UploadedFile
from src.shared.decorators.require_abilities_decorator import require_abilities
from src.shared.decorators.ignore_abilities_decorator import ignore_abilities
from src.modules.role.domain.enums.ability_codes_enum import AbilityCodes
from src.modules.role.domain.enums.ability_actions_enum import AbilityActions
from src.shared.domain.dto.pagination_response_dto import PaginationResponseDTO
from src.shared.types.user_request import UserRequest
from src.modules.user.domain.schemas.user_schema import (
    UserSchema,
)
from src.modules.user.domain.dto.user_find_many_dto import (
    UserFindManyDTO,
)
from src.modules.user.services.user_find_many_service import (
    UserFindManyService,
)
from src.modules.user.services.user_find_one_service import (
    UserFindOneService,
)
from src.modules.user.domain.dto.user_create_dto import (
    UserCreateDTO,
)
from src.modules.user.services.user_create_service import (
    UserCreateService,
)
from src.modules.user.domain.dto.user_update_dto import (
    UserUpdateDTO,
)
from src.modules.user.services.user_update_service import (
    UserUpdateService,
)
from src.modules.user.domain.dto.user_update_status_dto import (
    UserUpdateStatusDTO,
)
from src.modules.user.services.user_update_status_service import (
    UserUpdateStatusService,
)
from src.modules.user.services.user_delete_service import (
    UserDeleteService,
)
from src.modules.user.services.user_update_avatar_service import (
    UserUpdateAvatarService,
)

from src.modules.user.services.user_delete_avatar_service import (
    UserDeleteAvatarService,
)
from src.modules.user.services.user_find_self_service import (
    UserFindSelfService,
)

from src.modules.user.services.user_update_self_service import (
    UserUpdateSelfService,
)
from src.modules.user.domain.dto.user_update_self_dto import UserUpdateSelfDTO

router = Router(tags=["Users"])


@router.get("", response=PaginationResponseDTO[UserSchema])
@require_abilities(AbilityCodes.USERS, AbilityActions.READ)
def find_all(request, dto: Query[UserFindManyDTO]):
    user_request: UserRequest = request.user
    return UserFindManyService.execute(dto.dict(), user_request)


@router.get("/self", response=UserSchema)
@ignore_abilities
def find_self(request):
    user_request: UserRequest = request.user
    return UserFindSelfService.execute(user_request)


@router.get("/{id}", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.READ)
def find_one(request, id: int):
    user_request: UserRequest = request.user
    return UserFindOneService.execute(id, user_request)


@router.post("", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.CREATE)
def create(request, dto: UserCreateDTO):
    user_request: UserRequest = request.user
    return UserCreateService.execute(dto, user_request)


@router.put("/self", response=UserSchema)
@ignore_abilities
def update_self(request, dto: UserUpdateSelfDTO):
    user_request: UserRequest = request.user
    return UserUpdateSelfService.execute(dto, user_request)


@router.put("/{id}", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.UPDATE)
def update(request, id: int, dto: UserUpdateDTO):
    user_request: UserRequest = request.user
    return UserUpdateService.execute(id, dto, user_request)


@router.patch("/{id}/status", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.UPDATE)
def update_status(request, id: int, dto: UserUpdateStatusDTO):
    user_request: UserRequest = request.user
    return UserUpdateStatusService.execute(id, dto, user_request)


@router.post("/{id}/avatar", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.UPDATE)
def update_avatar(request, id: int, file: UploadedFile = File(...)):
    user_request: UserRequest = request.user
    return UserUpdateAvatarService.execute(id, file, user_request)


@router.delete("/{id}/avatar", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.DELETE)
def delete_avatar(request, id: int):
    user_request: UserRequest = request.user
    return UserDeleteAvatarService.execute(id, user_request)


@router.delete("/{id}", response=UserSchema)
@require_abilities(AbilityCodes.USERS, AbilityActions.DELETE)
def delete(request, id: int):
    user_request: UserRequest = request.user
    return UserDeleteService.execute(id, user_request)
